src/bridge/MemberListSyncer.ts

```diff
@@ -301,36 +307,34 @@ export class MemberListSyncer {
 
         log.debug("Got %s matrix join events to inject.", entries.length);
         this.usersToJoin = entries.length;
-        const d = promiseutil.defer();
-        // take the first entry and inject a join event
-        const joinNextUser = () => {
-            const entry = entries.shift();
-            if (!entry) {
-                d.resolve();
-                return;
-            }
+
+        let entry: MemberJoinEntry|undefined;
+        // eslint-disable-next-line no-cond-assign
+        while (entry = entries.shift()) {
             this.usersToJoin--;
             if (entry.userId.startsWith("@-")) {
-                joinNextUser();
-                return;
+                // Ignore guest users.
+                continue;
             }
             log.debug(
                 "Injecting join event for %s in %s (%s left) is_frontier=%s",
                 entry.userId, entry.roomId, entries.length, entry.frontier
             );
-            Bluebird.cast(injectJoinFn(entry.roomId, entry.userId, entry.displayName, entry.frontier)).timeout(
-                this.server.getMemberListFloodDelayMs()
-            ).then(() => {
-                joinNextUser();
-            }).catch(() => {
+            try {
+                const freshConnection =
+                    await injectJoinFn(entry.roomId, entry.userId, entry.displayName, entry.frontier);
+
+                // We only need to delay if the connection is "fresh" i.e. makes an impact on the ircd.
+                // If this is just reestablising a connection to the proxy then we can go as fast as
+                // we like!
+                if (freshConnection) {
+                    await promiseutil.delay(this.server.getMemberListFloodDelayMs());
```

Comment:
These are joins *to* the IRCD from Matrix, and for cold starts this will impact the ircd (hence the need for a delay).

My previous version of the code cocked up though and delayed all connections for a few MS, when actually we should be racing (and doing these concurrently).